Editor: allow using property grid to set Preferences

ericoporto opened the pull request:
![property_grid_preferences](https://user-images.githubusercontent.com/2244442/150656808-b4f6801a-6100-45cb-8c5f-892ceb22c144.gif)

This change modifies the Editor Preferences, it adds a tabbed panel, and allows modifying the settings using a Property Grid. For now all properties available in the Advanced tab are available elsewhere. The idea is that should we need to add a preference, we can add there more easily.

The way it works, is the EditorPreferences has some properties that are hidden using `[Browseable(false)]`, we also copy it when the preferences window is loaded. Any changes modifies this copy. When the user click ok, the values of the properties of this now modified copy is applied to the actual app settings (preferences).

I am not too sure on the names of the tab panels.

Fix #1477

ivan-mogilko commented:
I can see this works, but really hope we could come to some plan to improve this code-wise at some point. There are two things that are irking me personally in the current code:
1. Have to clone AppSettings object and then copy it back. AppSettings is a complicated class with underlying mechanisms and potential side-effects (from AppSettingsBase / SettingBase parent classes). Idk if i am too paranoid over this.
2. Have to list the cloned properties by hand. This means if one adds a new property they would have to remember to also copy it back and forth in Clone and Apply respectively.

OTOH, while searching in the web, I've seen few examples where people would also pass AppSettings into property grid, so *maybe* p1 is not dangerous...

Regarding p2, is it possible to use reflection instead?

---

By the way, I found that it is possible to "filter" properties on a grid using any custom attribute(s), not only "Browseable".
There's a property called [PropertyGrid.BrowsableAttributes](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.propertygrid.browsableattributes?view=windowsdesktop-6.0). This potentially allows to even split the properties into several grids (e.g. if you want to place these on separate tabs).

---

Strictly speaking about bugs:
1. "Color Theme" throws an unhandled exception if you type something wrong there. Actually it does not provide any list to choose from, so idk how to use it.
2. Neither of the path properties allow to browse for files, you may only type paths by hand.

ericoporto commented:
- > Have to clone AppSettings object and then copy it back.

I don't copy it, I copy it's properties. Reading the Docs I too could not figure out if copying the whole class could have any side-effects. I didn't found any through testing, but eventually decided it was more stress-free to just copy the properties.

- > is it possible to use reflection and loop over properties instead?

I think I did it now, I made that the attributes that are reassigned have to be browsable just to make sure we are only applying things set through the property grid.

- > There's a feature called PropertyGrid.BrowsableAttributes. This potentially allows to even split the properties into several grids

Thanks, that is really useful, I ended up using some info from there to make up the previously mentioned implementation.

---

- >Strictly speaking about bugs:
  >
  >"Color Theme" throws an unhandled exception if you type something wrong there. Actually it does not provide any list to choose from, so idk how to use it.

Thanks! I had not noticed that. I added now a list there, using a type converter. It's here: https://github.com/adventuregamestudio/ags/pull/1510/commits/4486710db1e459447923c2af5f3d63c29a631bf7

I have a problem that I could not figure where to place the `ColorThemeTypeConverter`, because it's relying in the `Factory.GUIController.ColorThemes.Themes` to give me the color themes, I don't think I should place in AGS Types, but I also don't know where to put it in the Editor side of the code. 

- > Neither of the path properties allow to browse for files, you may only type paths by hand.

Ah, now added a directory (new game, import) and a file browser (paint program).

- > The enums may benefit from description attributes, to have human-readable value strings.

Added those, based on the text that was already in use in the previous GUI.

ivan-mogilko commented:
This seem to work well now. Unless someone has further notes, I believe we may merge this.

ericoporto commented:
@persn , is this ok for you too?

ericoporto commented:
I think because it modifies a pane I need to update the color themes to account for this change. Not sure. I also can't seem to find where the color themes are stored.

persn commented:
I got busy and lost track of this whole feature, but I can try and force some time in my schedule to check the code

ericoporto commented:
That's fine, when you find time is good. :)

persn commented on Editor/AGS.Editor/Entities/EditorPreferences.cs:
```diff
@@ -242,6 +298,37 @@ private bool GetSettingsFromRegistry()
+               .ToList();
+        }
+
+        public AppSettings CloneAppSettings()
```
I don't get what the purpose of this is? I haven't read the entire conversation all the way back to the issue mind you, but ideally I shouldn't have to go anywhere outside of the code I'm looking at to understand it in the first place.

What confuses me is that I have to assume that we clone `AppSettings`  so that we can undo any changes made by a user if they hit cancel. But this use case is already covered in `ApplicationSettingsBase` with the `Reset()` and `Save()` isn't it?

ericoporto replied:
I don't know, [`ApplicationSettingsBase.Reset()`](https://docs.microsoft.com/en-us/dotnet/api/system.configuration.applicationsettingsbase.reset?view=netframework-4.6.2) has the following text:
> Restores the persisted application settings values to their corresponding default properties.

Is this wrong and it actually reloads the setting?

persn commented:
> I think because it modifies a pane I need to update the color themes (the JSON files) to account for this change. Not sure. I also can't seem to find where the JSON color themes are stored in AGS repository (Dark, DarkGrey, ...). Did we never integrated them in our builds/installer/...?

When I made feature PR with the color themes I suggested we made a repo or something where people could upload json and maintain them but I don't think anything came of it ever. The goto place for the latest updated theme is probably with the release itself

![image](https://user-images.githubusercontent.com/1510803/152229168-07139d65-d931-412c-8d55-1b9aa218f8b1.png)

ericoporto commented on Editor/AGS.Editor/Entities/EditorPreferences.cs:
```diff
@@ -242,6 +298,37 @@ private bool GetSettingsFromRegistry()
             return success;
         }
 
+        private List<PropertyInfo> GetBrowsableProperties()
```
Thanks! Lazy evaluation is a concept I just recently became familiar with, it's a good information knowing that's how LINQ works! I just adjusted for this!

ericoporto commented on Editor/AGS.Editor/Entities/EditorPreferences.cs:
```diff
@@ -3,55 +3,95 @@
+        {
+            Dictionary<string, string> colorThemeList = new Dictionary<string, string>();
+
+            foreach (ColorTheme colorTheme in Factory.GUIController.ColorThemes.Themes)
```
That is a much shorter code I didn't knew was possible! Thanks! Adjusted!

ericoporto commented:
I think it would be ok to add these here in the repo itself, in a new directory Themes under the Editor folder. In AGS4 the theme format will be simplified (in some PR from Alan), so if it was added elsewhere, it will need at least two directories anyhow.

ivan-mogilko commented:
> I think it would be ok to add these here in the repo itself, in a new directory Themes under the Editor folder. @ivan-mogilko , you ok with this approach?

In regards to use of themes, I don't have much of an opinion, so long as it works and is consistent.

ericoporto commented:
I have been thinking about going ahead with this version as is and if someone can improve later, it's fine, but I really need to have somewhere in the preferences to put the keystore related info for the Android build.

ivan-mogilko commented:
Okay, this does not affect serialization format and seem safe enough to merge now and adjust later if necessary.
